File: artifacts/api-server/src/routes/modules.ts
// Rutas de modulos: listar y obtener modulos educativos por ano de bachillerato
// Los modulos representan los 3 anos de la especialidad de informatica
import { Router } from "express";
import { db } from "@workspace/db";
import { modulosTable, nivelesTable, retosTable } from "@workspace/db";
import { eq, and } from "drizzle-orm";
import { requireAuth } from "../lib/auth";

const router = Router();

// GET /modules - Obtener todos los modulos activos (soporta filtro ?anio=N)
router.get("/modules", requireAuth, async (req, res) => {
  try {
    const anio = req.query["anio"] ? Number(req.query["anio"]) : null;

    let query = db.select().from(modulosTable).where(eq(modulosTable.activo, true)).$dynamic();

    if (anio) {
      query = db.select().from(modulosTable).where(and(eq(modulosTable.activo, true), eq(modulosTable.anio_bachillerato, anio))).$dynamic();
    }

    const modulos = await query.orderBy(modulosTable.anio_bachillerato);
    res.json(modulos);
  } catch (err) {
    req.log.error({ err }, "Error al obtener modulos");
    res.status(500).json({ error: "server_error", message: "Error interno" });
  }
});

// GET /modules/:moduleId - Obtener un modulo especifico por ID
router.get("/modules/:moduleId", requireAuth, async (req, res) => {
  try {
    const moduleId = Number(req.params.moduleId);
    const [modulo] = await db
      .select()
      .from(modulosTable)
      .where(eq(modulosTable.id, moduleId));

    if (!modulo) {
      res.status(404).json({ error: "not_found", message: "Modulo no encontrado" });
      return;
    }

    res.json(modulo);
  } catch (err) {
    req.log.error({ err }, "Error al obtener modulo");
    res.status(500).json({ error: "server_error", message: "Error interno" });
  }
});

// GET /modules/:moduleId/levels - Obtener los niveles de un modulo
router.get("/modules/:moduleId/levels", requireAuth, async (req, res) => {
  try {
    const moduleId = Number(req.params.moduleId);
    const niveles = await db
      .select()
      .from(nivelesTable)
      .where(and(eq(nivelesTable.id_modulo, moduleId), eq(nivelesTable.activo, true)))
      .orderBy(nivelesTable.orden);

    res.json(niveles);
  } catch (err) {
    req.log.error({ err }, "Error al obtener niveles del modulo");
    res.status(500).json({ error: "server_error", message: "Error interno" });
  }
});

// GET /modules/:moduleId/challenges - Obtener los retos de un modulo
router.get("/modules/:moduleId/challenges", requireAuth, async (req, res) => {
  try {
    const moduleId = Number(req.params.moduleId);

    // Obtiene los retos del modulo con informacion del nivel
    const retos = await db
      .select({
        id: retosTable.id,
        nombre: retosTable.nombre,
        descripcion: retosTable.descripcion,
        tipo_juego: retosTable.tipo_juego,
        id_modulo: retosTable.id_modulo,
        id_nivel: retosTable.id_nivel,
        tiempo_limite: retosTable.tiempo_limite,
        puntos_maximos: retosTable.puntos_maximos,
        numero_preguntas: retosTable.numero_preguntas,
        activo: retosTable.activo,
        creado_en: retosTable.creado_en,
      })
      .from(retosTable)
      .where(and(eq(retosTable.id_modulo, moduleId), eq(retosTable.activo, true)));

    // Para cada reto, agrega informacion del nivel y modulo
    const retosConInfo = await Promise.all(
      retos.map(async (reto) => {
        const [nivel] = await db
          .select()
          .from(nivelesTable)
          .where(eq(nivelesTable.id, reto.id_nivel));

        return { ...reto, nivel: nivel || null };
      })
    );

    res.json(retosConInfo);
  } catch (err) {
    req.log.error({ err }, "Error al obtener retos del modulo");
    res.status(500).json({ error: "server_error", message: "Error interno" });
  }
});

export default router;

File: artifacts/api-server-python/routers/modules.py
from fastapi import APIRouter, Depends, HTTPException
from db import get_db, fetchone, fetchall
from auth_utils import get_current_user

router = APIRouter()


def serialize_row(row):
    if row is None:
        return None
    from datetime import datetime
    r = dict(row)
    for k, v in r.items():
        if isinstance(v, datetime):
            r[k] = v.isoformat()
    return r


@router.get("/modules")
def get_modules(anio: int = None, user=Depends(get_current_user)):
    with get_db() as conn:
        if anio:
            rows = fetchall(conn, "SELECT * FROM modulos WHERE activo = 1 AND anio_bachillerato = %s ORDER BY anio_bachillerato", (anio,))
        else:
            rows = fetchall(conn, "SELECT * FROM modulos WHERE activo = 1 ORDER BY anio_bachillerato")
    return [serialize_row(r) for r in rows]


@router.get("/modules/{module_id}")
def get_module(module_id: int, user=Depends(get_current_user)):
    with get_db() as conn:
        modulo = fetchone(conn, "SELECT * FROM modulos WHERE id = %s", (module_id,))
    if not modulo:
        raise HTTPException(404, "Modulo no encontrado")
    return serialize_row(modulo)


@router.get("/modules/{module_id}/levels")
def get_module_levels(module_id: int, user=Depends(get_current_user)):
    with get_db() as conn:
        rows = fetchall(conn, "SELECT * FROM niveles WHERE id_modulo = %s AND activo = 1 ORDER BY orden", (module_id,))
    return [serialize_row(r) for r in rows]


@router.get("/modules/{module_id}/challenges")
def get_module_challenges(module_id: int, user=Depends(get_current_user)):
    with get_db() as conn:
        retos = fetchall(conn, "SELECT * FROM retos WHERE id_modulo = %s AND activo = 1", (module_id,))
        result = []
        for reto in retos:
            reto = dict(reto)
            nivel = fetchone(conn, "SELECT * FROM niveles WHERE id = %s", (reto["id_nivel"],))
            reto["nivel"] = serialize_row(nivel)
            result.append(serialize_row(reto))
    return result
